ethereum/permissioning/src/main/java/tech/pegasys/pantheon/ethereum/permissioning/TransactionSmartContractPermissioningController.java

```diff
@@ -0,0 +1,176 @@
+ * transactionAllowed(address,address,uint256,uint256,uint256,bytes)
+ */
+public class TransactionSmartContractPermissioningController
+    implements TransactionPermissioningProvider {
+  private final Address contractAddress;
+  private final TransactionSimulator transactionSimulator;
+
+  // full function signature for connection allowed call
+  private static final String FUNCTION_SIGNATURE =
+      "transactionAllowed(address,address,uint256,uint256,uint256,bytes)";
+  // hashed function signature for connection allowed call
+  private static final BytesValue FUNCTION_SIGNATURE_HASH = hashSignature(FUNCTION_SIGNATURE);
+
+  // The first 4 bytes of the hash of the full textual signature of the function is used in
+  // contract calls to determine the function being called
+  private static BytesValue hashSignature(final String signature) {
+    return Hash.keccak256(BytesValue.of(signature.getBytes(UTF_8))).slice(0, 4);
+  }
+
+  // True from a contract is 1 filled to 32 bytes
+  private static final BytesValue TRUE_RESPONSE =
+      BytesValue.fromHexString(
+          "0x0000000000000000000000000000000000000000000000000000000000000001");
+  private static final BytesValue FALSE_RESPONSE =
+      BytesValue.fromHexString(
+          "0x0000000000000000000000000000000000000000000000000000000000000000");
+
+  /**
+   * Creates a permissioning controller attached to a blockchain
+   *
+   * @param contractAddress The address at which the permissioning smart contract resides
+   * @param transactionSimulator A transaction simulator with attached blockchain and world state
+   */
+  public TransactionSmartContractPermissioningController(
+      final Address contractAddress, final TransactionSimulator transactionSimulator) {
+    this.contractAddress = contractAddress;
+    this.transactionSimulator = transactionSimulator;
+  }
+
+  /**
+   * Check whether a given transaction should be permitted for the current head
+   *
+   * @param transaction The transaction to be examined
+   * @return boolean of whether or not to permit the connection to occur
+   */
+  @Override
+  public boolean isPermitted(final Transaction transaction) {
+    final BytesValue payload = createPayload(transaction);
+    final CallParameter callParams =
+        new CallParameter(null, contractAddress, -1, null, null, payload);
+
+    final Optional<Boolean> contractExists =
+        transactionSimulator.doesAddressExistAtHead(contractAddress);
+
+    if (contractExists.isPresent() && !contractExists.get()) {
+      throw new IllegalStateException("Transaction permissioning contract does not exist");
+    }
+
+    final Optional<TransactionSimulatorResult> result =
+        transactionSimulator.processAtHead(callParams);
+
+    if (result.isPresent()) {
+      switch (result.get().getResult().getStatus()) {
+        case INVALID:
+          throw new IllegalStateException(
+              "Transaction permissioning transaction found to be Invalid");
```

Comment:
We should probably change this to `"Transaction permissioning call found to be Invalid"` to avoid misunderstandings.